### DISCUSSION TITLE
Extracting text contained inside a rect using quads

### DISCUSSION THREAD

**Discussion by andrescase** [pymupdf/PyMuPDF#1340] [Looking for help] [answered]
Hi,

I was wondering if there is any way to extract any words contained within the area of a rect object or using coordinates in a PDF file.

Thank you in advance for your help,

**Comment by andrescase** [comment]
Was able to solve it by looking for words within coordinate interval.

**Reply by JorjMcKie** [reply]
I would do this:
* assume your rectangle is named `rect`
* make a list of words on the page: `words = page.get_text("words")`.
* each item of the word list has 4 starting floats following by the word string: `(x0, y0, x1, y1, "word",...)` followed by some more info

```python
for word in words:
    bbox = fitz.Rect(word[:4])  # make bbox from subtuple
    if bbox in rect:
        print("word '%s' in rectangle" % w[4])
```

**Reply by JorjMcKie** [reply]
If you need to check if a word is contained in a ``quad``, and not necessarily in a rectangle, things are currently more complicated.
* First you have to map the quad such that its sides are parallel to the x- resp. y-axis. If the quad is more general, i.e. has no corners with 90° angles, then there is no solution at all, **_currently_**. But you can check whether a quad is congruent to a rectangle via `quad.is_rectangular`.
* There are ways to compute a matrix such that the line from `quad.ul` to `quad.ur` lies on the x-axis. This matrix (let us call it `mat`) thus converts the complete quad to a traditional rectangle.
* Then check whether `bbox * mat` is in the rectangle of the converted quad.

```python
mat = fitz.planish_line(quad.ul, quad.ur)  # compute the conversion matrix
rect = (quad * mat).rect  # rect of converted quad

# ...
for word in words:
    bbox = fitz.Rect(word[:4])  # make bbox from subtuple
    if bbox * mat in rect:
        print("word '%s' in quad" % w[4])
```

A future version of PyMuPDF will allow to conveniently check, whether a point is inside a quad. Then you would only need to check whether all bbox corners are inside the quad.

**Comment by JorjMcKie** [comment]
Oops - sorry: I forgot I already have that said function included 🙄 - you **_can_** do this:

```python
for word in words:
    bbox = fitz.Rect(word[:4])  # make bbox from subtuple
    if bbox in quad:
        print("word '%s' in quad" % w[4])
```

**Reply by JorjMcKie** [reply]
All of the following checks **_are possible_**:
* ``point in quad``
* ``rect in quad``
* ``quad in quad``

**Comment by JorjMcKie** [comment, marked as answer]
I have just updated the documentation: you will find my previous comments in the section for class ``Quad``.
